garage/tf/regressors/categorical_mlp_regressor_with_model.py

```diff
@@ -0,0 +1,255 @@
+        hidden_b_init (callable): Initializer function for the bias
+            of intermediate dense layer(s). The function should return a
+            tf.Tensor. Default is zero initializer.
+        output_nonlinearity (callable): Activation function for output dense
+            layer. It should return a tf.Tensor. Set it to None to
+            maintain a softmax activation.
+        output_w_init (callable): Initializer function for the weight
+            of output dense layer(s). The function should return a
+            tf.Tensor. Default is Glorot uniform initializer.
+        output_b_init (callable): Initializer function for the bias
+            of output dense layer(s). The function should return a
+            tf.Tensor. Default is zero initializer.
+        optimizer (tf.Optimizer): Optimizer for minimizing the negative
+            log-likelihood.
+        optimizer_args (dict): Arguments for the optimizer. Default is None,
+            which means no arguments.
+        tr_optimizer (tf.Optimizer): Optimizer for trust region approximation.
+        tr_optimizer_args (dict): Arguments for the trust region optimizer.
+            Default is None, which means no arguments.
+        use_trust_region (bool): Whether to use trust region constraint.
+        max_kl_step (float): KL divergence constraint for each iteration.
+        normalize_inputs (bool): Bool for normalizing inputs or not.
+        layer_normalization (bool): Bool for using layer normalization or not.
+        no_initial_trust_region (bool): Bool to specify if trust region
+            constraint has been used initially.
+    """
+
+    def __init__(self,
+                 input_shape,
+                 output_dim,
+                 name='CategoricalMLPRegressorWithModel',
+                 hidden_sizes=(32, 32),
+                 hidden_nonlinearity=tf.nn.tanh,
+                 hidden_w_init=tf.glorot_uniform_initializer(),
+                 hidden_b_init=tf.zeros_initializer(),
+                 output_nonlinearity=tf.nn.softmax,
+                 output_w_init=tf.glorot_uniform_initializer(),
+                 output_b_init=tf.zeros_initializer(),
+                 optimizer=None,
+                 optimizer_args=None,
+                 tr_optimizer=None,
+                 tr_optimizer_args=None,
+                 use_trust_region=True,
+                 max_kl_step=0.01,
+                 normalize_inputs=True,
+                 layer_normalization=False,
+                 no_initial_trust_region=True):
+
+        super().__init__(input_shape, output_dim, name)
+        self._use_trust_region = use_trust_region
+        self._no_initial_trust_region = no_initial_trust_region
+        self._max_kl_step = max_kl_step
+        self._normalize_inputs = normalize_inputs
+        self._first_optimized = not no_initial_trust_region
+
+        with tf.variable_scope(
+                self._name, reuse=False) as self._variable_scope:
```

Comment:
Avoid assigning the return of the context manager to an instance attribute. Do
```python
with tf.variable_scope(self.name, reuse=False) as vs:
    self._variable_scope = vs
```
Please also update the same thing for `ContinuousMLPRegressorWithModel` and `GaussianMLPRegressorWithModel`.